File: iosApp/Conservatio/Theme/Localization.swift
import Foundation
import SwiftUI

/// Conservatio's lightweight string catalog. Mirrors the web `i18n.ts`
/// keys exactly so the same translation table can be reused across
/// platforms.
///
/// Wired with a single global `t("…")` helper plus a `LanguagePreference`
/// observable that backs Settings. Default language is English; the user
/// flips to Greek from Settings and the choice is persisted in
/// UserDefaults under `app.language`.

@MainActor
@Observable
final class LanguagePreference {
    static let shared = LanguagePreference()

    var code: String {
        didSet { UserDefaults.standard.set(code, forKey: "app.language") }
    }

    private init() {
        code = UserDefaults.standard.string(forKey: "app.language") ?? "en"
    }
}

@MainActor
func t(_ key: String) -> String {
    let lang = LanguagePreference.shared.code
    return (Strings.table[lang]?[key]) ?? (Strings.table["en"]?[key]) ?? key
}

private enum Strings {
    static let table: [String: [String: String]] = [
        "en": en,
        "el": el,
    ]

    static let en: [String: String] = [
        // Tabs
        "nav.dashboard": "Home",
        "nav.objects": "Objects",
        "nav.projects": "Projects",
        "nav.clients": "Clients",
        "nav.reports": "Reports",
        "nav.settings": "Settings",
        // Login
        "login.continueGoogle": "Continue with Google",
        "login.continueApple": "Continue with Apple",
        "login.continueGitHub": "Continue with GitHub",
        "login.finishingOauth": "Finishing sign-in…",
        "login.errSignIn": "Sign-in failed. Try again.",
        // Dashboard
        "dash.welcome": "Welcome back",
        "dash.title": "Conservatio",
        "dash.intro":
            "Document heritage objects, write condition reports, manage conservation projects and clients.",
        "dash.signedInTip":
            "You're signed in. Every new record is uploaded to your account.",
        "dash.offlineTip":
            "Offline mode. Records live on this device only.",
        "dash.newObject": "New Object",
        "dash.takePhoto": "Take Photo",
        "dash.newReport": "New Report",
        "dash.newProject": "New Project",
        "dash.recentObjects": "Recent Objects",
        // Stat
        "stat.objects": "Objects",
        "stat.reports": "Reports",
        "stat.projects": "Projects",
        "stat.clients": "Clients",
        // Generic
        "g.save": "Save",
        "g.cancel": "Cancel",
        "g.edit": "Edit",
        "g.delete": "Delete",
        "g.signOut": "Sign Out",
        "g.signIn": "Sign In",
        "g.syncNow": "Sync Now",
        "g.notSet": "Not set",
        // Settings
        "settings.title": "Settings",
        "settings.language": "Language",
        "settings.account": "Account",
        "settings.profile": "Profile",
        "settings.displayName": "Display name",
        "settings.signOut": "Sign Out",
        "settings.dangerTitle": "Local data",
        "settings.clearAll": "Clear all data",
        "settings.loadSample": "Load sample data",
        "settings.clearConfirm": "This will remove every object, client, project, and report. Are you sure?",
        // Cloud
        "cloud.title": "Cloud Storage",
        "cloud.subtitle": "Where your conservation records and photos live. Connect a provider to add capacity beyond the free Conservatio quota.",
        "cloud.conservatio": "Conservatio Cloud",
        "cloud.conservatioFreeNote": "Included with every account. 2 GB free, then upgradeable.",
        "cloud.signedOutUsage": "Sign in to see your cloud usage.",
        "cloud.usageLabel": "Storage used",
        "cloud.providers": "Other providers",
        "cloud.comingSoon": "Coming soon",
        "cloud.providersHint": "Connecting a third-party provider lets Conservatio mirror your records and photos there. None are wired yet; tap to express interest.",
        "cloud.connectGoogleDrive": "Google Drive",
        "cloud.connectICloud": "iCloud",
        "cloud.connectOneDrive": "OneDrive",
        "cloud.connectMega": "MEGA",
        "cloud.connectDropbox": "Dropbox",
        "cloud.interestNoted": "Noted",
        "cloud.connected": "Connected",
        "g.offline": "Offline",
    ]

    static let el: [String: String] = [
        // Tabs
        "nav.dashboard": "Αρχική",
        "nav.objects": "Αντικείμενα",
        "nav.projects": "Έργα",
        "nav.clients": "Πελάτες",
        "nav.reports": "Αναφορές",
        "nav.settings": "Ρυθμίσεις",
        // Login
        "login.continueGoogle": "Συνέχεια με Google",
        "login.continueApple": "Συνέχεια με Apple",
        "login.continueGitHub": "Συνέχεια με GitHub",
        "login.finishingOauth": "Ολοκλήρωση σύνδεσης…",
        "login.errSignIn": "Αποτυχία σύνδεσης. Δοκίμασε ξανά.",
        // Dashboard
        "dash.welcome": "Καλώς ήρθατε",
        "dash.title": "Conservatio",
        "dash.intro":
            "Τεκμηρίωση αντικειμένων, αναφορές κατάστασης, διαχείριση έργων και πελατών.",
        "dash.signedInTip":
            "Είσαι συνδεδεμένος. Κάθε νέα καταχώρηση ανεβαίνει στον λογαριασμό σου.",
        "dash.offlineTip":
            "Λειτουργία εκτός σύνδεσης. Οι καταχωρήσεις μένουν μόνο σε αυτή τη συσκευή.",
        "dash.newObject": "Νέο αντικείμενο",
        "dash.takePhoto": "Φωτογραφία",
        "dash.newReport": "Νέα αναφορά",
        "dash.newProject": "Νέο έργο",
        "dash.recentObjects": "Πρόσφατα αντικείμενα",
        // Stat
        "stat.objects": "Αντικείμενα",
        "stat.reports": "Αναφορές",
        "stat.projects": "Έργα",
        "stat.clients": "Πελάτες",
        // Generic
        "g.save": "Αποθήκευση",
        "g.cancel": "Άκυρο",
        "g.edit": "Επεξεργασία",
        "g.delete": "Διαγραφή",
        "g.signOut": "Αποσύνδεση",
        "g.signIn": "Σύνδεση",
        "g.syncNow": "Συγχρονισμός",
        "g.notSet": "Δεν έχει οριστεί",
        // Settings
        "settings.title": "Ρυθμίσεις",
        "settings.language": "Γλώσσα",
        "settings.account": "Λογαριασμός",
        "settings.profile": "Προφίλ",
        "settings.displayName": "Εμφανιζόμενο όνομα",
        "settings.signOut": "Αποσύνδεση",
        "settings.dangerTitle": "Τοπικά δεδομένα",
        "settings.clearAll": "Διαγραφή όλων",
        "settings.loadSample": "Φόρτωση δείγματος",
        "settings.clearConfirm": "Θα αφαιρεθούν όλα τα αντικείμενα, πελάτες, έργα και αναφορές. Είστε σίγουροι;",
        // Cloud
        "cloud.title": "Αποθήκευση στο cloud",
        "cloud.subtitle": "Εδώ ζουν οι καταχωρήσεις και οι φωτογραφίες σας. Συνδέστε πάροχο για χώρο πέρα από τη δωρεάν ποσόστωση.",
        "cloud.conservatio": "Conservatio Cloud",
        "cloud.conservatioFreeNote": "Περιλαμβάνεται σε κάθε λογαριασμό. 2 GB δωρεάν, μετά αναβαθμίσιμο.",
        "cloud.signedOutUsage": "Συνδέσου για να δεις τη χρήση του cloud.",
        "cloud.usageLabel": "Χρήση αποθηκευτικού χώρου",
        "cloud.providers": "Άλλοι πάροχοι",
        "cloud.comingSoon": "Σύντομα",
        "cloud.providersHint": "Η σύνδεση τρίτου παρόχου επιτρέπει στο Conservatio να αντικατοπτρίζει τα αρχεία σας εκεί. Πατήστε για να εκδηλώσετε ενδιαφέρον.",
        "cloud.connectGoogleDrive": "Google Drive",
        "cloud.connectICloud": "iCloud",
        "cloud.connectOneDrive": "OneDrive",
        "cloud.connectMega": "MEGA",
        "cloud.connectDropbox": "Dropbox",
        "cloud.interestNoted": "Σημειώθηκε",
        "cloud.connected": "Συνδεδεμένο",
        "g.offline": "Εκτός σύνδεσης",
    ]
}

File: androidApp/src/main/java/com/conservatio/android/ui/Localization.kt
package com.conservatio.android.ui

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext

/**
 * Conservatio's lightweight string catalog for Android. Mirrors the web
 * `i18n.ts` and iOS `Localization.swift` so all three platforms share the
 * same translation keys.
 *
 * Stored in SharedPreferences under "app.language". Toggle via
 * Settings; default is English.
 */
object Strings {
    fun current(context: Context): String =
        context.getSharedPreferences("conservatio", Context.MODE_PRIVATE)
            .getString("app.language", "en") ?: "en"

    fun set(context: Context, code: String) {
        context.getSharedPreferences("conservatio", Context.MODE_PRIVATE)
            .edit()
            .putString("app.language", code)
            .apply()
    }

    fun t(code: String, key: String): String =
        table[code]?.get(key) ?: table["en"]?.get(key) ?: key

    private val en = mapOf(
        // Tabs
        "nav.dashboard" to "Home",
        "nav.objects" to "Objects",
        "nav.projects" to "Projects",
        "nav.clients" to "Clients",
        "nav.reports" to "Reports",
        "nav.settings" to "Settings",
        // Login
        "login.continueGoogle" to "Continue with Google",
        "login.continueApple" to "Continue with Apple",
        "login.continueGitHub" to "Continue with GitHub",
        "login.finishingOauth" to "Finishing sign-in…",
        "login.errSignIn" to "Sign-in failed. Try again.",
        // Dashboard
        "dash.welcome" to "Welcome back",
        "dash.title" to "Conservatio",
        "dash.intro" to
            "Document heritage objects, write condition reports, manage conservation projects and clients.",
        "dash.signedInTip" to
            "You're signed in. Every new record is uploaded to your account.",
        "dash.offlineTip" to
            "Offline mode. Records live on this device only.",
        "dash.newObject" to "New Object",
        "dash.takePhoto" to "Take Photo",
        "dash.newReport" to "New Report",
        "dash.newProject" to "New Project",
        "dash.recentObjects" to "Recent Objects",
        // Stat
        "stat.objects" to "Objects",
        "stat.reports" to "Reports",
        "stat.projects" to "Projects",
        "stat.clients" to "Clients",
        // Generic
        "g.save" to "Save",
        "g.cancel" to "Cancel",
        "g.edit" to "Edit",
        "g.delete" to "Delete",
        "g.signOut" to "Sign Out",
        "g.signIn" to "Sign In",
        "g.syncNow" to "Sync Now",
        "g.notSet" to "Not set",
        // Settings
        "settings.title" to "Settings",
        "settings.language" to "Language",
        "settings.account" to "Account",
        "settings.profile" to "Profile",
        "settings.displayName" to "Display name",
        "settings.signOut" to "Sign Out",
        "settings.dangerTitle" to "Local data",
        "settings.clearAll" to "Clear all data",
        "settings.loadSample" to "Load sample data",
    )

    private val el = mapOf(
        // Tabs
        "nav.dashboard" to "Αρχική",
        "nav.objects" to "Αντικείμενα",
        "nav.projects" to "Έργα",
        "nav.clients" to "Πελάτες",
        "nav.reports" to "Αναφορές",
        "nav.settings" to "Ρυθμίσεις",
        // Login
        "login.continueGoogle" to "Συνέχεια με Google",
        "login.continueApple" to "Συνέχεια με Apple",
        "login.continueGitHub" to "Συνέχεια με GitHub",
        "login.finishingOauth" to "Ολοκλήρωση σύνδεσης…",
        "login.errSignIn" to "Αποτυχία σύνδεσης. Δοκίμασε ξανά.",
        // Dashboard
        "dash.welcome" to "Καλώς ήρθατε",
        "dash.title" to "Conservatio",
        "dash.intro" to
            "Τεκμηρίωση αντικειμένων, αναφορές κατάστασης, διαχείριση έργων και πελατών.",
        "dash.signedInTip" to
            "Είσαι συνδεδεμένος. Κάθε νέα καταχώρηση ανεβαίνει στον λογαριασμό σου.",
        "dash.offlineTip" to
            "Λειτουργία εκτός σύνδεσης. Οι καταχωρήσεις μένουν μόνο σε αυτή τη συσκευή.",
        "dash.newObject" to "Νέο αντικείμενο",
        "dash.takePhoto" to "Φωτογραφία",
        "dash.newReport" to "Νέα αναφορά",
        "dash.newProject" to "Νέο έργο",
        "dash.recentObjects" to "Πρόσφατα αντικείμενα",
        // Stat
        "stat.objects" to "Αντικείμενα",
        "stat.reports" to "Αναφορές",
        "stat.projects" to "Έργα",
        "stat.clients" to "Πελάτες",
        // Generic
        "g.save" to "Αποθήκευση",
        "g.cancel" to "Άκυρο",
        "g.edit" to "Επεξεργασία",
        "g.delete" to "Διαγραφή",
        "g.signOut" to "Αποσύνδεση",
        "g.signIn" to "Σύνδεση",
        "g.syncNow" to "Συγχρονισμός",
        "g.notSet" to "Δεν έχει οριστεί",
        // Settings
        "settings.title" to "Ρυθμίσεις",
        "settings.language" to "Γλώσσα",
        "settings.account" to "Λογαριασμός",
        "settings.profile" to "Προφίλ",
        "settings.displayName" to "Εμφανιζόμενο όνομα",
        "settings.signOut" to "Αποσύνδεση",
        "settings.dangerTitle" to "Τοπικά δεδομένα",
        "settings.clearAll" to "Διαγραφή όλων",
        "settings.loadSample" to "Φόρτωση δείγματος",
    )

    private val table: Map<String, Map<String, String>> = mapOf(
        "en" to en,
        "el" to el,
    )
}

/** CompositionLocal exposing the active language to Composable trees. */
val LocalLanguageCode = compositionLocalOf { "en" }

/** Read a translated string from inside a Composable. */
@Composable
fun str(key: String): String {
    val code = LocalLanguageCode.current
    return Strings.t(code, key)
}

/**
 * Wrap your app's root with this to make `str("…")` available everywhere.
 * Re-reads the saved language on each composition so toggles in Settings
 * propagate immediately.
 */
@Composable
fun WithLanguage(content: @Composable () -> Unit) {
    val context = LocalContext.current
    val code by remember { mutableStateOf(Strings.current(context)) }
    CompositionLocalProvider(LocalLanguageCode provides code) {
        content()
    }
}
